### ISSUE TITLE
Implementing Box Developer Edition

### ISSUE THREAD

**Issue by bgoldman** [box/box-java-sdk#149]
CODE CHANGES

Existing class BoxApiConnection:

1. Create new public static method getAppAuthConnection: Args for enterpriseId, clientId, clientSecret, privateKey, privateKeyPassword. Requests enterprise token using new authenticate method. Instantiates and returns new BoxApiConnection.
2. Create new public static method getAppUserConnection: Args for userId, clientId, clientSecret, privateKey, privateKeyPassword. Requests user token using new authenticate method. Instantiates and returns new BoxApiConnection.
3. Create new version of public authenticate: Args for entityId, type, private key. Creates a JWT assertion and then uses that assertion to request an auth token. Called by getAppAuthConnection and getAppUserConnection.
4. Create new private static method readPrivateKey: Args for privateKey and privateKeyPassword. Converts the private key string and returns PrivateKey instance, decrypting if necessary. Called by authenticate.

Existing class BoxUser:

1. Create new public method createAppUser: Same args as createEnterpriseUser. Calls createEnterpriseUser, setting is_platform_access_only=true.
2. Modify existing public method createEnterpriseAccessUser: Add param for isPlatformAccessOnly.

NOTES

1. Creating an app user will be done by pass-through using BoxUser.createEnterpriseUser, without needing to modify this method beyond adding one param. It will authenticate using a BoxApiConnection instance created from getAppAuthConnection.

2. Getting an app user will be done by instantiating a new BoxUser or by using BoxUser.getCurrentUser(), without needing to modify either of these methods. It will authenticate using a BoxApiConnection instance created from getAppUserConnection.

3. Getting an app user’s info will be done using BoxUser.getInfo, without needing to modify this method. It will authenticate using a BoxApiConnection instance created from getAppUserConnection.

4. Editing an app user will be done using BoxUser.updateInfo, without needing to modify this method. It will authenticate using a BoxApiConnection instance created from getAppUserConnection.

5. Deleting an app user will be done using BoxUser.delete, without needing to modify this method. It will authenticate using a BoxApiConnection instance created from getAppUserConnection.

**Comment by pflory** [comment]
Hi Bgoldman,

Are these enhancement requests and/or do you have code for any of these enhancements?  

Two of these methods (getAppAuthConnection and getAppUserConnection) sound helpful for a problem that surfaced about 2 weeks ago that "broke" our Oauth code that simulates a user logging into box to authenticate (oauth) to initiate the BoxApiConnection. (The java code hits the web login page/form and works through the Grant/authorize sequence, but after 3 years no longer works)

From communication with Box Support, it seems the Oauth "handshake" was modified about 2 weeks ago, and now requires different cookies, but the details have not been documented or figured out yet.

Did you happen to encounter this problem and/or find a workaround?


Regards,
Paul

**Comment by bgoldman** [comment]
I was hired on contract by Box to make updates to the Java and .NET SDKs, specifically to implement Box Developer Edition, to add functionality for App Auth and App Users. I was asked to submit a review of my proposed changes as a github issue prior to writing code.

I'm glad to hear that the two primary methods here sound helpful for you!

Let me know if you have any other feedback or questions.

**Comment by pflory** [comment]
Thanks for the info!  Is there a (tentative) timeline for implementing some or all of this?

Our client is in a bind since the Oauth process changed recently, mentioned earlier, breaking their backend integration between their web CMS and Box, so curious if an "official" API might be coming soon ... or workarounds will be needed in the meantime.

**Comment by bgoldman** [comment]
Pull requests coming soon, within the coming weeks :)

**Comment by pflory** [comment]
Hi BGoldman,

Will your Pull Request happen to help out in our particular situation described here?

In order to "bootstrap" the Box connection, our java code uses the boxOAuthLogin method (shown below) so that no human interaction is required for our backend integration to connect to Box.com. However, as mentioned earlier, something changed in the Oauth authentication process, and this method now fails (after working for ~ 3 years).

Would any of your new PR features address this kind of "bootstrapping" for authentication?

Hopefully this request is clear... if not, would be happy to clarify.

Regards,
Paul

Initial bootstrapping for boxAPIConnection and its authenticate() method:
```java
            	String authCode = boxOAuthLogin(boxOAuthUser, boxOAuthUserPassword, docBoxClientID, docBoxClientSecret, "http://localhost:4000");
            	boxAPIConnection = new BoxAPIConnection(docBoxClientID, docBoxClientSecret);
            	boxAPIConnection.authenticate(authCode);

```

Our code, below, for mimicking a human user stepping through the OAuth web forms to get the authCode for the boxAPIConnection.authenticate() method (shown above):

```java
	private String boxOAuthLogin(String username, String password, String clientId, String clientSecret, String redirectUri) throws IntegrationException {
		log.debug("Inside boxOAuthLogin method");

		String authCode = "";
		try {
			HttpClient httpClient = new HttpClient();

			// Get login page
			String postUri = BOX_COM_API_OAUTH2_AUTHORIZE_URL + "?response_type=code&client_id=" + clientId + "&redirect_uri="
					+ URLEncoder.encode(redirectUri, "UTF-8"); 
			GetMethod getMethod = new GetMethod(postUri);

			int statusCode = httpClient.executeMethod(getMethod);
			
			if(statusCode == 200) {

				String response = getMethod.getResponseBodyAsString();
				int begin = response.indexOf("request_token") + 17;
				int end = response.indexOf(";var site_root") - 1;
	
				String requestToken = null;
	
				if (begin != -1) {
					requestToken = response.substring(begin, end);
				}
	
				// Submit user and password
				PostMethod postMethod = new PostMethod(BOX_COM_API_OAUTH2_AUTHORIZE_URL);
	
				postMethod.addRequestHeader("Connection", "keep-alive");
			
				postMethod.addRequestHeader("Referer", postUri);
	
				postMethod.addParameter("__login", "1");
				postMethod.addParameter("_pw_sql", "");
				postMethod.addParameter("_redirect_url", "/api/oauth2/authorize" + REFERER_PARAMS);
				postMethod.addParameter("client_id", clientId);
				postMethod.addParameter("dologin", "1");
				postMethod.addParameter("folder", "");
				postMethod.addParameter("folder_id", "");
				postMethod.addParameter("login", username);
				postMethod.addParameter("login_or_register_mode", "login");
				postMethod.addParameter("login_submit", "Authorizing...");
				postMethod.addParameter("new_login_or_register_mod...", "");
				postMethod.addParameter("password", password);
				postMethod.addParameter("redirect_uri", redirectUri);
				postMethod.addParameter("reg_step", "");
				postMethod.addParameter("request_token", requestToken);
				postMethod.addParameter("response_type", "code");
				postMethod.addParameter("scope", "root_readwrite");
				postMethod.addParameter("state", "authenticated");
				postMethod.addParameter("submit1", "1");
	
				httpClient.executeMethod(postMethod);
				requestToken = response.substring(begin, end);
				String ic = getIc(postMethod.getResponseBodyAsString());
	
				// Get Box authorization code
				postUri = BOX_COM_API_OAUTH2_AUTHORIZE_URL;
	
				postMethod = new PostMethod(postUri);
				postMethod.addRequestHeader("Connection", "keep-alive");
				postMethod.addRequestHeader("Host", "app.box.com");
				postUri = BOX_COM_API_OAUTH2_AUTHORIZE_URL + "?response_type=code&client_id=" + clientId + "&redirect_uri="
						+ URLEncoder.encode(redirectUri, "UTF-8");
				postMethod.addRequestHeader("Referer", postUri);
	
				postMethod.addParameter("client_id", clientId);
				postMethod.addParameter("consent_accept", "Grant access to Box");
				postMethod.addParameter("doconsent", "doconsent");
				postMethod.addParameter("folder_id", "");
				postMethod.addParameter("ic", ic);
				postMethod.addParameter("redirect_uri", redirectUri);
				postMethod.addParameter("request_token", requestToken);
				postMethod.addParameter("response_type", "code");
				postMethod.addParameter("scope", "root_readwrite");
				postMethod.addParameter("state", "authenticated");
	
				httpClient.executeMethod(postMethod);
				int httpStatusCode = postMethod.getStatusCode();
				log.debug("HTTP STATUS2: " + httpStatusCode + " Response Body: " + postMethod.getResponseBodyAsString() + " Header-Location: "
						+ postMethod.getResponseHeader("Location"));
				authCode = getCode(postMethod.getResponseHeader("Location").toString());
				log.debug("authCode: " + authCode);
			} else if(statusCode == 407) {
				throw new IntegrationException("Require Proxy credentials to access box.com");
			}

		} catch (IOException ioe) {
			log.error("error during oauth login to box.com {} " , ioe.getMessage());
			throw new IntegrationException("error during oauth login to box.com");
		}

		return authCode;
	}

```

**Comment by gcurtis** [comment]
@bgoldman I really like the overall direction of these changes. I just have a few additional suggestions around auth and method signatures.

Instead of changing the `BoxAPIConnection` class, it might be a better idea to subclass it. While I'd like to avoid an inheritance hierarchy, I think a lot of existing behavior is going to have to change. The most obvious example is around token refreshing. The `canRefresh()`, `needsRefresh()`, and `refresh()` methods are all going to have completely different behavior with app auth. For example:

```java
public class BoxAppAuthConnection extends BoxAPIConnection {

	private BoxAppAuthConnection(subID, subType, clientID, clientSecret,
		privateKey, privateKeyPassword) {

		super(clientID, clientSecret);
	}

	public static BoxAppAuthConnection createEnterpriseConnection(enterpriseID,
		clientID, clientSecret, privateKey, privateKeyPassword) {

		return new BoxAppAuthConnection(enterpriseID, "enterprise", clientID,
			clientSecret, privateKey, privateKeyPassword);
	}

	public static BoxAppAuthConnection createUserConnection(enterpriseID,
		clientID, clientSecret, privateKey, privateKeyPassword) {
		
		return new BoxAppAuthConnection(enterpriseID, "user", clientID,
			clientSecret, privateKey, privateKeyPassword);
	}

	@Override
	public boolean canRefresh() { }

	@Override
	public boolean needsRefresh() { }

	@Override
	public void refresh() { }

	// Any other methods that need to be overridden.
}
```

The other suggestion is around method signature changes.

> Modify existing public method createEnterpriseAccessUser: Add param for isPlatformAccessOnly.

Unfortunately we can't change the signature of any public methods for compatibility reasons. Instead, you'll need to create a new private method for `createEnterpriseUser(boolean isPlatformAccessOnly)` and then have the public methods call that.

Everything else looks great! I really appreciate the thought that went into this design. Let me know what you think of the changes.

**Comment by gcurtis** [comment]
@bgoldman - one other thought. The SDK tries to keep dependencies down to a minimum so as to avoid bloat and to make it easier for apps that consume the SDK (right now the only dependency is a small JSON parser). The logic for building the JWT assertion seems pretty complicated, so it's understandable if we need to pull in another library, but it'd be preferable if it's small and doesn't contain many transient dependencies.

**Comment by gcurtis** [comment]
@pflory this feature is related to Box Developer Edition which is currently in beta. You can find more information on what Box Developer Edition is in the [API documentation](https://box-content.readme.io/v2.0/docs/box-developer-edition).

This new way of authenticating does bypass user interaction, but it doesn't necessarily grant you the same privileges as normal OAuth. Specifically, it grants your app access to its own "enterprise", but not a user's personal account. You can contact Box Support to get more information on whether or not this might be helpfulf for your use-case.

**Comment by pflory** [comment]
Thanks, @gcurtis. We were aware of that Beta SDK, but because this is ("was" since it's broke) a live, production integration with hundreds of customers using the Box documents, the Dev team decided to go with the Box SDK until the Developer Edition was in a post-beta status. 

If anyone else knows what changes happened to the Oauth process/flow, and can offer a fix/workaround/hack in the meantime, it would be greatly appreciated.

Paul

**Comment by gcurtis** [comment]
@pflory unfortunately there really isn't a good solution for what you're trying to do. Mimicking user interaction to get an auth code may sometimes work, but it can also break at any time (as it has in your case). If you do manage to figure out what changed with the login flow, there still aren't any guarantees that it won't change again in a few weeks/months.

The only workarounds are to figure out what changed, update your code, and risk the login page changing again. Or you can bootstrap the process with a real user to get an access token. Once you have an access/refresh token pair, you'll never need an auth code again (provided you use the refresh token at least once a month).

**Comment by bgoldman** [comment]
@gcurtis Glad you like the direction of these changes.

Regarding subclassing vs not subclassing, I understand your point, but I think it could still turn out well without subclassing - although during implementation, I might change my mind. I propose an idea: how about I build it without subclassing, and if you don't like how it turns out, I move it into a subclass?

Regarding `createEnterpriseUser`, I think there was some miscommunication. I don't intend to add an arg, I intend to add

**Comment by gcurtis** [comment]
@bgoldman yeah I'm completely fine with trying it without subclassing it first. Adding `isPlatformAccessOnly` to `CreateUserParams` makes a lot more sense, so that sounds good too.

I look forward to seeing the pull request!

**Comment by bgoldman** [comment]
@gcurtis thoughts? https://github.com/box/box-java-sdk/pull/151

**Comment by gcurtis** [comment]
This has been implemented as of a89780872bf86db83f06ebbb55adbebcd9fd15ae.
